Potential Cleanup: thread safety annotations for custom mutex wrapping std::mutex

### Feature description

This is low priority.

https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#warning-flags

With a recent copy of clang, warnings pop up with -Wthread-safety-analysis. There may be more, but I build a very limited set of drivers:

- frmts/vrt/vrtdataset.cpp
- frmts/zarr/zarr_v2_array.cpp
- frmts/zarr/zarr_v3_array.cpp

The code may well be fine.

e.g.

```
third_party/gdal/frmts/vrt/vrtdataset.cpp:3260:18: error: mutex 'poMutex' is not held on every path through here [-Werror,-Wthread-safety-analysis]
 3260 |     auto oIter = oMap.find(osKey);
      |                  ^
third_party/gdal/frmts/vrt/vrtdataset.cpp:3259:18: note: mutex acquired here
 3259 |         poMutex->lock();
      |                  ^
third_party/gdal/frmts/vrt/vrtdataset.cpp:3265:18: error: releasing mutex 'poMutex' that was not held [-Werror,-Wthread-safety-analysis]
 3265 |         poMutex->unlock();
      |                  ^
third_party/gdal/frmts/vrt/vrtdataset.cpp:3277:19: error: mutex 'poMutex' is not held on every path through here [-Werror,-Wthread-safety-analysis]
 3277 |     oMap[osKey] = poDS;
      |                   ^
third_party/gdal/frmts/vrt/vrtdataset.cpp:3276:18: note: mutex acquired here
 3276 |         poMutex->lock();
      |                  ^
third_party/gdal/frmts/vrt/vrtdataset.cpp:3279:18: error: releasing mutex 'poMutex' that was not held [-Werror,-Wthread-safety-analysis]
 3279 |         poMutex->unlock();
```

and

```
third_party/gdal/frmts/zarr/zarr_v3_array.cpp:398:32: error: mutex 'm_oMutex' is not held on every path through here [-Werror,-Wthread-safety-analysis]
  398 |     auto poTilePresenceArray = OpenTilePresenceCache(false);
      |                                ^
third_party/gdal/frmts/zarr/zarr_v3_array.cpp:397:18: note: mutex acquired here
  397 |         m_oMutex.lock();
      |                  ^
third_party/gdal/frmts/zarr/zarr_v3_array.cpp:417:26: error: releasing mutex 'm_oMutex' that was not held [-Werror,-Wthread-safety-analysis]
  417 |                 m_oMutex.unlock();
      |                          ^
third_party/gdal/frmts/zarr/zarr_v3_array.cpp:425:18: error: releasing mutex 'm_oMutex' that was not held [-Werror,-Wthread-safety-analysis]
  425 |         m_oMutex.unlock();
```


### Additional context

_No response_

--- a/cmake/helpers/GdalCompilationFlags.cmake
+++ b/cmake/helpers/GdalCompilationFlags.cmake
@@ -123,6 +123,9 @@ else ()
   detect_and_set_cxx_warning_flag(deprecated-copy-dtor)
   detect_and_set_cxx_warning_flag(implicit-fallthrough)
   detect_and_set_cxx_warning_flag(weak-vtables)
+  if(DEFINED ENV{CI})
+      detect_and_set_cxx_warning_flag(thread-safety-analysis)
+  endif()
 
   # Not sure about the minimum version, but clang 12 complains about \file, @cond Doxygen_Suppress, etc.
   if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND "${CMAKE_CXX_COMPILER_VERSION}" VERSION_GREATER_EQUAL 18.0.0)

--- a/frmts/vrt/vrtdataset.cpp
+++ b/frmts/vrt/vrtdataset.cpp
@@ -3250,33 +3250,44 @@ void VRTDataset::ClearStatistics()
 }
 
 /************************************************************************/
-/*                   VRTMapSharedResources::Get()                       */
+/*                 VRTMapSharedResources::LockGuard()                   */
 /************************************************************************/
 
-GDALDataset *VRTMapSharedResources::Get(const std::string &osKey) const
+std::unique_ptr<std::lock_guard<std::mutex>>
+VRTMapSharedResources::LockGuard() const
 {
-    if (poMutex)
-        poMutex->lock();
-    auto oIter = oMap.find(osKey);
-    GDALDataset *poRet = nullptr;
-    if (oIter != oMap.end())
-        poRet = oIter->second;
-    if (poMutex)
-        poMutex->unlock();
-    return poRet;
+    std::unique_ptr<std::lock_guard<std::mutex>> poLockGuard;
+    if (m_bUseMutex)
+    {
+        poLockGuard = std::make_unique<std::lock_guard<std::mutex>>(m_oMutex);
+    }
+    return poLockGuard;
 }
 
 /************************************************************************/
 /*                   VRTMapSharedResources::Get()                       */
 /************************************************************************/
 
+GDALDataset *VRTMapSharedResources::Get(const std::string &osKey) const
+{
+    auto poLockGuard = LockGuard();
+    CPL_IGNORE_RET_VAL(poLockGuard);
+    auto oIter = m_oMap.find(osKey);
+    GDALDataset *poRet = nullptr;
+    if (oIter != m_oMap.end())
+        poRet = oIter->second;
+    return poRet;
+}
+
+/************************************************************************/
+/*                   VRTMapSharedResources::Insert()                    */
+/************************************************************************/
+
 void VRTMapSharedResources::Insert(const std::string &osKey, GDALDataset *poDS)
 {
-    if (poMutex)
-        poMutex->lock();
-    oMap[osKey] = poDS;
-    if (poMutex)
-        poMutex->unlock();
+    auto poLockGuard = LockGuard();
+    CPL_IGNORE_RET_VAL(poLockGuard);
+    m_oMap[osKey] = poDS;
 }
 
 /************************************************************************/
@@ -3285,7 +3296,7 @@ void VRTMapSharedResources::Insert(const std::string &osKey, GDALDataset *poDS)
 
 void VRTMapSharedResources::InitMutex()
 {
-    poMutex = &oMutex;
+    m_bUseMutex = true;
 }
 
 /*! @endcond */

--- a/frmts/vrt/vrtdataset.h
+++ b/frmts/vrt/vrtdataset.h
@@ -113,9 +113,11 @@ class CPL_DLL VRTMapSharedResources
     void InitMutex();
 
   private:
-    std::mutex oMutex{};
-    std::mutex *poMutex = nullptr;
-    std::map<std::string, GDALDataset *> oMap{};
+    mutable std::mutex m_oMutex{};
+    bool m_bUseMutex = false;
+    std::map<std::string, GDALDataset *> m_oMap{};
+
+    std::unique_ptr<std::lock_guard<std::mutex>> LockGuard() const;
 
     CPL_DISALLOW_COPY_ASSIGN(VRTMapSharedResources)
 };

--- a/frmts/zarr/zarr_v2_array.cpp
+++ b/frmts/zarr/zarr_v2_array.cpp
@@ -442,8 +442,13 @@ bool ZarrV2Array::LoadTileData(const uint64_t *tileIndices, bool bUseMutex,
                      ->GetStreamingFilename(osFilename);
 
     // First if we have a tile presence cache, check tile presence from it
+    std::unique_ptr<std::lock_guard<std::mutex>> poLockGuard;
     if (bUseMutex)
-        m_oMutex.lock();
+    {
+        poLockGuard = std::make_unique<std::lock_guard<std::mutex>>(m_oMutex);
+    }
+    CPL_IGNORE_RET_VAL(poLockGuard);
+
     auto poTilePresenceArray = OpenTilePresenceCache(false);
     if (poTilePresenceArray)
     {
@@ -462,16 +467,13 @@ bool ZarrV2Array::LoadTileData(const uint64_t *tileIndices, bool bUseMutex,
                                       eByteDT, &byValue) &&
             byValue == 0)
         {
-            if (bUseMutex)
-                m_oMutex.unlock();
             CPLDebugOnly(ZARR_DEBUG_KEY, "Tile %s missing (=nodata)",
                          osFilename.c_str());
             bMissingTileOut = true;
             return true;
         }
     }
-    if (bUseMutex)
-        m_oMutex.unlock();
+    poLockGuard.reset();
 
     VSILFILE *fp = nullptr;
     // This is the number of files returned in a S3 directory listing operation

--- a/frmts/zarr/zarr_v3_array.cpp
+++ b/frmts/zarr/zarr_v3_array.cpp
@@ -393,8 +393,13 @@ bool ZarrV3Array::LoadTileData(const uint64_t *tileIndices, bool bUseMutex,
                      ->GetStreamingFilename(osFilename);
 
     // First if we have a tile presence cache, check tile presence from it
+    std::unique_ptr<std::lock_guard<std::mutex>> poLockGuard;
     if (bUseMutex)
-        m_oMutex.lock();
+    {
+        poLockGuard = std::make_unique<std::lock_guard<std::mutex>>(m_oMutex);
+    }
+    CPL_IGNORE_RET_VAL(poLockGuard);
+
     auto poTilePresenceArray = OpenTilePresenceCache(false);
     if (poTilePresenceArray)
     {
@@ -413,16 +418,13 @@ bool ZarrV3Array::LoadTileData(const uint64_t *tileIndices, bool bUseMutex,
                                       eByteDT, &byValue) &&
             byValue == 0)
         {
-            if (bUseMutex)
-                m_oMutex.unlock();
             CPLDebugOnly(ZARR_DEBUG_KEY, "Tile %s missing (=nodata)",
                          osFilename.c_str());
             bMissingTileOut = true;
             return true;
         }
     }
-    if (bUseMutex)
-        m_oMutex.unlock();
+    poLockGuard.reset();
 
     VSILFILE *fp = nullptr;
     // This is the number of files returned in a S3 directory listing operation

